README.md
# Home automation at the tips of your fingers through Facebook messenger
[![Circle CI](https://circleci.com/gh/Eaton-Software/messenger-home.svg?style=svg)](https://circleci.com/gh/Eaton-Software/messenger-home)
[![Circle CI](https://circleci.com/gh/Eaton-Software/messenger-home/tree/develop.svg?style=svg)](https://circleci.com/gh/Eaton-Software/messenger-home/tree/develop)
[![Build Status](https://travis-ci.org/Eaton-Software/messenger-home.svg?branch=develop)](https://travis-ci.org/Eaton-Software/messenger-home)
[![bitHound Overall Score](https://www.bithound.io/github/Eaton-Software/messenger-home/badges/score.svg)](https://www.bithound.io/github/Eaton-Software/messenger-home)
[![bitHound Dependencies](https://www.bithound.io/github/Eaton-Software/messenger-home/badges/dependencies.svg)](https://www.bithound.io/github/Eaton-Software/messenger-home/master/dependencies/npm)

A bot created with the Facebook messenger API that helps automate your home.

----
## Functionalities
- Retrieving indoor temperature
  - Using an MPU-6050 sensor hooked up to an Arduino Uno.
- Monitoring security cameras
  - Multiple security camera capability through webcam.js.
- Notifying events
  - Simulated door bell using a button connected to an Arduino Uno with an accompanying LED.
- Light conversation
  - It's polite!

----
## Built With
- node.js
- express.js
- ejs
- webcam.js
- arduino uno
- johnny-five
- MPU-6050



CHANGELOG.md
## Changelog Version 0.2.0
Continuous integration, code linters, oh my!

## Changelog Version 0.1.0



test/test.js
var request = require('supertest');

describe('loading express', function () {
  var server;
  before(function () {
    server = require('../bin/www');
  });
  after(function (done) {
    server.close(done);
  });
  it('responds to /bot/cam', function testSlash(done) {
    request(server)
    .get('/bot/cam')
    .expect(200, done);
  });
  it('responds to webhooks', function testVerify(done) {
    request(server)
    .get('/bot/webhook')
    .query({'hub.verify_token': 'eleven_eaton_pl'})
    .query({'hub.challenge': 'give_me_this'})
    .expect('give_me_this', done);
  });
});


test/.eslintrc.js
module.exports = {
  'env': {
    mocha: true
  }
}



app.js
'use strict';
var express = require('express');
var path = require('path');
// var favicon = require('serve-favicon');
var winston = require('winston');
var expressWinston = require('express-winston');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');

var fb = require('./routes/fb');

var app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');

// uncomment after placing your favicon in /public
//app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
var logger = new winston.Logger({
  transports: [
    new winston.transports.File({
      level: 'info',
      filename: './debug-logs.log',
      handleExceptions: true,
      json: true,
      maxsize: 5242880, //5MB
      maxFiles: 5,
      colorize: false
    }),
    new winston.transports.Console({
      level: 'debug',
      handleExceptions: true,
      json: false,
      colorize: true
    })
  ],
  exitOnError: false
});

app.use(expressWinston.logger({
  winstonInstance: logger,
  meta: false,
  expressFormat: true
}));

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use('/bot', fb);

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  var err = new Error('Not Found');
  err.status = 404;
  next(err);
});

// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
  app.use(function(err, req, res, next) { // eslint-disable-line no-unused-vars
    res.status(err.status || 500);
    res.render('error', {
      message: err.message,
      error: err
    });
  });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) { // eslint-disable-line no-unused-vars
  res.status(err.status || 500);
  res.render('error', {
    message: err.message,
    error: {}
  });
});


module.exports = app;


gulpfile.js
'use strict';
var gulp = require('gulp');
var mocha = require('gulp-mocha');
var eslint = require('gulp-eslint');

gulp.task('test', function() {
  return gulp.src('./test/test.js', {read: false})
  .pipe(mocha());
});

gulp.task('lint', function () {
  // ESLint ignores files with "node_modules" paths.
  // So, it's best to have gulp ignore the directory as well.
  // Also, Be sure to return the stream from the task;
  // Otherwise, the task may end before the stream has finished.
  return gulp.src(['**/*.js','!node_modules/**', '!public/**'])
  // eslint() attaches the lint output to the "eslint" property
  // of the file object so it can be used by other modules.
  .pipe(eslint())
  // eslint.format() outputs the lint results to the console.
  // Alternatively use eslint.formatEach() (see Docs).
  .pipe(eslint.format());
  // To have the process exit with an error code (1) on
  // lint error, return the stream and pipe to failAfterError last.
  // .pipe(eslint.failAfterError());
});


routes/fb.js
'use strict';
var express = require('express');
var router = express.Router();
var request = require('request');
var keys = require('../keys');
var s = require('../res/strings/english');  //  Strings
var multer  = require('multer');
var upload = multer({dest: 'uploads/'});
var fs = require('fs');
var logger = require('winston');


// Index route
router.get('/', function (req, res) {
  res.send('Hello, I am a bot!');
});

// cam view
router.get('/cam/', function (req, res) {
  res.render('camSelect');
});

// cam view
router.get('/cam/:num', function (req, res) {
  res.render('cam', { num: req.params.num });
});

router.post('/pic/:num', upload.single('webcam'), function(req, res) {
  var num = req.params.num;
  fs.rename(req.file.path, 'public/shot' + num + '.jpg', function(err) {
    if(err) {
      res.status(500).send(err);
    } else {
      res.send('Screenshot success!');
    }
  });
});

// for Facebook verification
router.get('/webhook/', function (req, res) {
  if ('hub.verify_token' in req.query && 'hub.challenge' in req.query && req.query['hub.verify_token'] === 'eleven_eaton_pl') {
    res.send(req.query['hub.challenge']);
  } else {
    res.status(400).send('Error, wrong token');
  }
});




/////////////////////////
// Messaging Functions //
/////////////////////////

function sendMessage(messageData, sender) {
  request({
    url: 'https://graph.facebook.com/v2.6/me/messages',
    qs: {access_token:token},
    method: 'POST',
    json: {
      recipient: {id:sender},
      message: messageData
    }
  }, function(error, response, body) {
    if (error) {
      logger.error('Error sending messages: ', error);
    } else if (response.body.error) {
      logger.error('Error: ', body.error);
    }
  });
}

function sendText(sender, text) {
  var messageData = {text:text};
  sendMessage(messageData, sender);
}

function sendCam(sender, cam) {
  logger.info('Sending image...');
  var messageData = {
    'attachment': {
      'type': 'image',
      'payload': {
        'url': 'https://fb.jagels.us/shot'+ cam +'.jpg'
      }
    }
  };
  sendMessage(messageData, sender);
}

function sendAllCams(sender, arr) {
  var elements = [];
  for(var cam in arr) {
    elements.push({
      'title' : 'Cam ' + arr[cam],
      'image_url': 'https://fb.jagels.us/shot' + arr[cam] + '.jpg'
    });
  }
  logger.info(elements);
  var messageData = {
    'attachment': {
      'type': 'template',
      'payload': {
        'template_type': 'generic',
        'elements': elements
      }
    }
  };
  sendMessage(messageData, sender);
}

var recTemp = 'null';
function sendTemperature(sender) {
  logger.info('Displaying temperature...');
  var messageData = {
    'attachment': {
      'type': 'template',
      'payload': {
        'template_type': 'generic',
        'elements': [{
          'title': 'Current Temperature',
          'subtitle': recTemp + ' °F'
        }]
      }
    }
  };
  sendMessage(messageData, sender);
}

function sendGeneric(sender) {
  logger.info('Displaying generic message...');
  var messageData = {
    'attachment': {
      'type': 'template',
      'payload': {
        'template_type': 'generic',
        'elements': [{
          'title': 'First card',
          'subtitle': 'Element #1 of an hscroll',
          'image_url': 'http://messengerdemo.parseapp.com/img/rift.png',
          'buttons': [
            {
              'type': 'web_url',
              'url': 'https://www.messenger.com',
              'title': 'web url'
            },
            {
              'type': 'postback',
              'title': 'Postback',
              'payload': 'Payload for first element in a generic bubble'
            }
          ]
        },
          {
            'title': 'Second card',
            'subtitle': 'Element #2 of an hscroll',
            'image_url': 'http://messengerdemo.parseapp.com/img/gearvr.png',
            'buttons': [{
              'type': 'postback',
              'title': 'Postback',
              'payload': 'Payload for second element in a generic bubble'
            }]
          }
        ]
      }
    }
  };
  sendMessage(messageData, sender);
}





/////////////////////////
// Language Processing //
/////////////////////////
function has(msg, sub) {
  var index = msg.toLowerCase().indexOf(sub);
  if(index >= 0) {
    var prev = msg.charAt(index-1);
    var after = msg.charAt(index+sub.length);
    if((prev==='' || prev===' ') && (after==='' || after===' ')){return true;}
  } return false;
}

function hasArr(msg, arr) {
  for(var sub in arr) {
    if(has(msg, arr[sub]))
      return true;
  }
  return false;
}

function assessPrompt(msg) {
  if(has(msg, 'help')) return 'help';
  if(hasArr(msg, s.words.CAM)) return 'cam';
  if(hasArr(msg, s.words.DISPLAY) && hasArr(msg, s.words.TEMPERATURE)) return 'temperature';
  if(hasArr(msg, s.words.STATE)) return 'state';
  if(hasArr(msg, s.words.GREET)) return 'greet';
  if(msg === 'Generic') return 'generic';
  return 'default';
}

var sender;
router.post('/webhook/', function (req, res) {
  var messaging_events = req.body.entry[0].messaging;
  for (var i = 0; i < messaging_events.length; i++) {
    var event = req.body.entry[0].messaging[i];
    sender = event.sender.id;
    if (event.message && event.message.text) {
      var text = event.message.text;
      logger.info('Received message: ' + text);
      switch(assessPrompt(text)) {
      case 'help':
        sendText(sender, s.responses.HELP);
        break;
      case 'generic':
        sendGeneric(sender);
        break;
      case 'temperature':
        sendTemperature(sender);
        break;
      case 'greet':
        sendText(sender, s.responses.GREET);
        break;
      case 'cam':
        var num = text.match(/\d+/);
        if(!num) {
          sendText(sender, 'Please specify a valid camera');
        } else {
          sendText(sender, 'Here\'s your picture!');
          if(has(text, 'all')) {
            sendAllCams(sender, num);
          } else {
            sendCam(sender, num[0]);
          }
        }
        break;
      case 'state':
        sendText(sender, s.responses.TFLUK);
        break;
      default:
        logger.info('No prompt recognized...');
      }
    }
    if (event.postback) {
      text = JSON.stringify(event.postback);
      sendText(sender, 'Postback received: '+text.substring(0, 200), token);
      continue;
    }
  }
  res.sendStatus(200);
});

router.post('/doorbell', function(req, res) {
  sendText(sender, 'Hey, there\'s someone at your door!');
  res.send('Rang doorbell...');
});

router.post('/getTemp', function(req, res) {
  recTemp = req.body.temp;
  res.send('Set temp...');
});

var token = keys.FB_ACCESS_TOKEN;
module.exports = router;
